### DISCUSSION TITLE
Clarification on Actions Auth

### DISCUSSION THREAD

**Discussion by aykaramba** [rabbibotton/clog#209] [General]
I am moving forward with my small gallery applications and trying to use the Clog Website Project template to learn how to setup routes with permissions.

I believe I understanding most aspects of how permissions work and it is all fairly straight forward.  The only part that is giving me problems is understanding what Actions Auth does in the menu.  Here is my sample code including the menu and the handler for the gallery menu item:

```
                      ; Menu         Menu Item         URL                  Handler     Actions Auth
(defparameter *menu* `(("Features" (("Home"            "/")
                                     ("Login"           "/login"             on-login    :login)
                                     ("Signup"          "/signup"            on-signup   :signup)
                                     ("Change Password" "/pass"              on-new-pass :change-password)
                                     ("Content"         "/content"           on-main     :content)
                                     ("Logout"          "/logout"            on-logout   :logout)))

                        ("Apps" (("Some web page"       "/content/some-web-page")
				 ("Gallery"             "/gallery"  on-gallery   :content)  
				 ("Logout"              "/logout"   on-logout    :logout)))
					       
		        ("Admin"   (("User List"        "/users"              on-users   :content)))
		        ("Help"    (("About"            "/content/about"))))
  "Setup website menu")

;; handler
(defun on-gallery (body)
  (init-site body)
  (create-web-page body
                        :content `(:menu ,*menu*
                           
			        :content ,(lambda (obj)
				          (create-div obj :content (random 23423432)) ;; testing text generation
				          (create-div obj :content (random 23423432)) ;; testing text generation
				          (test obj) ;; function prints some text on page
				          (form-upload obj) ;; function creates file upload form
				          (blah obj))) ;; function creates a 3 columnt gallery
		         :authorize t))
```
Basically I am looking to have Apps / Gallery menu item accessible only to logged in users. Authorize t makes sure /gallery is only accessible to authenticated users, that makes sense.  What I don't fully underand is the Actions Auth from above and how that relates to who can access / see the menu item.  I was under the impression that  Actions Auth was something like a group / user profile and that I could create arbitrary ones as necessary, however, it looks like the number of possible Actions Auth entries are tied to the Clog Website Project template and there are only a fixed number of profiles that can be used.

I have read the Auth Objects in the manual and related docs but I cannot find out information on how Actions Auth works.

Does anyone know?  Any tips or hints?

Thanks.

**Comment by rabbibotton** [comment]
```
(defun on-about (body)
  (init-site body)
  (create-web-page body :about `(:menu    ,*menu*
                                 :content "About Me")))
```
So :about while passed and _can_ be used by the template, is more of a ID for the page type.

```
(add-authorization '(:guest) '(:login :signup :main))
(add-authorization '(:member) '(:logout :main))
(add-authorization '(:admin) '(:users))
```

Is where you set up relationships between user levels and page type IDs

If no authorizations are made all users have access, as in :about

:users (the edit users in system page) can only be access by someone assigned the role of :admin

The association from user to profiles in the tutorial 31 is in init-site
```
 :roles (if profile
                                (if (equalp "admin"
                                            (getf profile :|username|))
                                    '(:member :admin)
                                    '(:member))
                                '(:guest))
```

If there is no profile set -:guest, if profile name is admin - is :member and :admin, otherwise is just a :member

**Comment by aykaramba** [comment]
All of that makes sense and it works great.  This is really straight forward and a really great way to quickly create an application with permissions.  This really makes me happy.

Just one point to any noobs following up, whenever you update the add-authorization form you will need to restart Clog for the updates to be usable / see the changes work.  That was the other permutation adding complexity to trying to figure out what worked, when and why.

Thank you again Rabbi.
